src/charm_state.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The module for managing charm state."""
+
+
+import logging
+import os
+import typing
+from dataclasses import dataclass
+
+import ops
+from pydantic import BaseModel, ValidationError
+
+import metadata
+
+# agent relation name
+AGENT_RELATION = "agent"
+
+logger = logging.getLogger()
+
+
+class Credentials(BaseModel):
+    """The credentials used to register to the Jenkins server.
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+def _get_jenkins_unit(
+    all_units: typing.Set[ops.Unit], current_app_name: str
+) -> typing.Optional[ops.Unit]:
+    """Get the Jenkins charm unit in a relation.
+
+    Args:
+        all_units: All units in a relation.
+        current_app_name: The Jenkins-agent application name.
+
+    Returns:
+        The Jenkins server application unit in the relation if found. None otherwise.
+    """
+    for unit in all_units:
+        # if the unit's application name is the same, this is peer unit. Otherwise, it is the
+        # Jenkins server unit.
+        if unit.app.name == current_app_name:
+            continue
+        return unit
+    return None
+
+
+def _get_credentials_from_agent_relation(
+    server_unit_databag: ops.RelationDataContent, unit_name: str
+) -> typing.Optional[Credentials]:
+    """Import server metadata from databag in agent relation.
+
+    Args:
+        server_unit_databag: The relation databag content from agent relation.
+        unit_name: The agent unit name.
+
+    Returns:
+        Metadata if complete values(url, secret) are set. None otherwise.
+    """
+    address = server_unit_databag.get("url")
+    secret = server_unit_databag.get(f"{unit_name}_secret")
+    if not address or not secret:
+        return None
```

Comment:
Isn't this an error? In this case there was relation data for the unit but it was not correct, i.e., it was missing either the address or secret. Shouldn't that mean the unit goes into error state saying that the data that was received over the relation was not valid